### DISCUSSION TITLE
Waku: Minimalistic React Framework

### DISCUSSION THREAD

**Discussion by dai-shi** [cyco130/vite-rsc#5] [Show and tell]
[Waku](https://github.com/dai-shi/waku) (formerly Wakuwork) is one of my projects to build a React framework. It's very opinionated about being minimal with several design decisions. We will see how it matches to the world in the coming years.

There must be a common part that I don't mind developing on my own and share it with other projects. That would probably be client bundling and server bundling. So, I look forward to this project.

Questions about Waku are welcome in this discussion threads.

**Comment by cyco130** [comment]
I did read a lot of your code while working on this! I'm sure will have tons of questions for you soon :)

**Reply by dai-shi** [reply]
thanks. the latest version is released within an hour: https://twitter.com/dai_shi/status/1642463300314333184
many internal things are changed since the previous version.

**Comment by dai-shi** [comment]
One of the difficulties I find is how to use `react-server-dom-webpack` with vite. Mostly hacked by now, but one of remaining things is how to use ESM loader with TypeScript, in Node.js. This seems hard in development mode where we want to transpile TS code on the fly as well as handling "use client/server" directives.

Solving this isn't my focus, but would need some experiments to move forward.

**Comment by cyco130** [comment]
Our plan for dev is to run a second Vite server with a separate cache directory, running in a separate process spawned with `node --condition react-server`, along with a require hook (already in `react-server-dom-webpack`) and an ESM loader (I have [experience dealing with them](https://github.com/cyco130/vavite/tree/main/packages/node-loader)) to transform external deps that have `use client`. We'll provide a bridge function that the SSR code can call to render server components.

For build, we'll just bundle the RSC build with ESBuild (SSR and client builds will work as usual).

Once we have something working and useful, we'll try to convince the Vite team to give us a third module graph so we can get rid of the workarounds :)

**Reply by dai-shi** [reply]
Cool. Looking forward to it. I will also experiment separate process or worker, but if there's a better support from vite, I'd definitely try to make mine migrate to it.
